uaclient/system/cpu_type.py

```diff
@@ -0,0 +1,187 @@
+        return "; ".join(
+            [
+                "{}:{}".format(key, value)
+                for key, value in sorted(fallback_dict.items())
+            ]
+        )
+
+    def _get_fallback_name(
+        self,
+        cpu_info,
+        fallback_fields,
+        re_pattern,
+    ):
+        return self._cpu_type_stringify(
+            fallback_dict={
+                field: self._get_field_value(
+                    cpu_info=cpu_info,
+                    re_pattern=re_pattern.format(field),
+                )
+                or ""
+                for field in fallback_fields
+            }
+        )
+
+
+class X86CPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/proc/cpuinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}\s+:\s+(?P<name>.+)"
+
+    @classmethod
+    def processor_match(cls, processor_type: str) -> bool:
+        return (
+            processor_type == "amd64"
+            or processor_type == "i386"
+            or processor_type == "i686"
+        )
+
+    def collect(self):
+        cpu_info = system.load_file(self.CPU_TYPE_FILE)
+        model_name = self._get_field_value(
+            cpu_info=cpu_info,
+            re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("model name"),
+        )
+
+        if model_name:
+            return model_name
+
+        # start backup plan if model name is not present in the data
+        return self._get_fallback_name(
+            cpu_info=cpu_info,
+            fallback_fields=("vendor_id", "cpu family", "model", "stepping"),
+            re_pattern=self.CPU_TYPE_REGEX_PATTERN,
+        )
+
+
+class PPCCPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/proc/cpuinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}\s+:\s+(?P<name>.+)"
+
+    @classmethod
+    def processor_match(cls, processor_type: str) -> bool:
+        return processor_type.startswith("ppc")
+
+    def collect(self):
+        cpu_info = system.load_file(self.CPU_TYPE_FILE)
+        return self._get_field_value(
+            cpu_info=cpu_info,
+            re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("cpu"),
+        )
+
+
+class S390CPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/proc/sysinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}:\s+(?P<name>.+)"
+
+    @classmethod
+    def processor_match(cls, processor_type: str) -> bool:
+        return processor_type.startswith("s390")
+
+    def collect(self):
+        cpu_info = system.load_file(self.CPU_TYPE_FILE)
+        cpu_type = (
+            self._get_field_value(
+                cpu_info=cpu_info,
+                re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("Type"),
+            )
+            or ""
+        )
+        cpu_model = (
+            self._get_field_value(
+                cpu_info=cpu_info,
+                re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("Model"),
+            )
+            or ""
+        )
+
+        return "Type:{}; Model:{}".format(cpu_type, cpu_model)
```

Comment:
Now here we can call
```suggestion
        return self._cpu_type_stringify({"type": cpu_type, "model": cpu_model})
```